nexus/src/app/certificate.rs

```diff
@@ -0,0 +1,230 @@
+                    direction: dropshot::PaginationOrder::Ascending,
+                    limit: std::num::NonZeroU32::new(1).unwrap(),
+                },
+            )
+            .await
+            .map_err(|e| {
+                Error::internal_error(&format!(
+                    "Failed to list certificates: {e}"
+                ))
+            })?;
+
+        let certificate = if let Some(certificate) = certs.get(0) {
+            certificate
+        } else {
+            return Ok(None);
+        };
+
+        Ok(Some(dropshot::ConfigTls::AsBytes {
+            certs: certificate.cert.clone(),
+            key: certificate.key.clone(),
+        }))
+    }
+
+    /// Refreshes the TLS configuration for the currently-running Nexus external
+    /// server. This involves either:
+    /// - Creating a new HTTPS server if once does not exist
+    /// - Refreshing an existing HTTPS server if it already exists
+    /// - Tearing down an HTTPS server if no certificates exist
+    pub async fn refresh_tls_config(
+        &self,
+        opctx: &OpContext,
+    ) -> Result<(), Error> {
+        let tls_config = self.get_nexus_tls_config(&opctx).await?;
+
+        let mut external_servers = self.external_servers.lock().await;
+
+        match (tls_config, external_servers.https.take()) {
+            // Create a new server, using server context from an existing HTTP
+            // server.
+            (Some(tls_config), None) => {
+                info!(self.log, "Refresh TLS: Creating HTTPS server");
+                let mut cfg = external_servers.config.clone();
+                cfg.bind_address.set_port(external_servers.https_port());
+                cfg.tls = Some(tls_config);
+
+                let context =
+                    external_servers.get_context().ok_or_else(|| {
+                        Error::internal_error("No server context available")
+                    })?;
+
+                let log =
+                    context.log.new(o!("component" => "dropshot_external"));
+                let server_starter_external = dropshot::HttpServerStarter::new(
+                    &cfg,
+                    crate::external_api::http_entrypoints::external_api(),
+                    context,
+                    &log,
+                )
+                .map_err(|e| {
+                    Error::internal_error(&format!(
+                        "Initializing HTTPS server: {e}"
+                    ))
+                })?;
+                external_servers.https.replace(server_starter_external.start());
+            }
+            // Refresh an existing server.
+            (Some(tls_config), Some(https)) => {
+                info!(
+                    self.log,
+                    "Refresh TLS: Refreshing HTTPS server at {}",
+                    https.local_addr()
+                );
+                https.refresh_tls(&tls_config).await.map_err(|e| {
+                    Error::internal_error(&format!("Cannot refresh TLS: {e}"))
```

Comment:
What state are we in at this point?  Is the server running with the previous set of certs?

Reply:
I'll unwrap explicitly here: 4bafeeefbcc14c81e7bc6852665d4bfb4b57e158 , this error should not happen